source/slang/slang-parameter-binding.cpp

```diff
@@ -329,11 +329,17 @@ struct UsedRanges
     }
 };
 
+namespace ParameterAndLayoutConstants
+{
+    const constexpr size_t kInputAttachmentIndexUnused = 0xFFFFFFFF;
+};
+
 struct ParameterBindingInfo
 {
     size_t              space = 0;
     size_t              index = 0;
     LayoutSize          count = 0;
+    size_t              inputAttachmentIndex = ParameterAndLayoutConstants::kInputAttachmentIndexUnused;
```

Comment:
We shouldn't be extending `ParameterBindingInfo` type here for something that is only used for a specific type of global var. What is this needed for? Can we just add a parameter cateogry for InputAttachment instead?

Reply:
input_attachment_index is part of the same conflict group as any other descriptor object (image binding 0, set 0, cannot be with a subpass binding 0, set 0). The difference is input_attachment_index is an additional conflict that has to be watched for.

Reply:
I still don't understand why we can't juse use `binding` for `inputAttachmentIndex`, in what situation can they diverge?

Reply:
_what each qualifier means:_

1. input_attachment_index references the index set when setting up renderpass's:
 ```
VkAttachmentReference renderPassAttachment{};
colorReference.attachment = 0;
```

2. binding & set is referencing the descriptor set and where we bound the renderpass's image view.

_why we need them both for error reporting:_
1. if we bound to `descriptor set 1, binding 0, attachment 0` and `descriptor set 0, binding 0, attachment 0`, we should warn since this is a conflict which [spec warns is illegal for input_attachment_index](https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_vulkan_glsl.txt).
```
layout (input_attachment_index = 0, set = 0, binding = 0) uniform subpassInput inputColor;
layout (input_attachment_index = 0, set = 1, binding = 0) uniform subpassInput inputColor2;
```

2. similarly, the same code here should conflict, but due to binding being the same, which is a problem since this is an impossible scenario to have (your descriptor is either for `input_attachment_index = 1` or `input_attachment_index = 0`, not both).
```
layout (input_attachment_index = 1, set = 0, binding = 0) uniform subpassInput inputColor;
layout (input_attachment_index = 0, set = 0, binding = 0) uniform subpassInput inputColor2;
```

3. finally, since this is a regular descriptor, the following code will also conflict: due to an impossible scenario of having two resources which are different using the same descriptor bindings.
```
layout (set = 0, binding = 0) uniform image1D inputTexture1D;
layout (input_attachment_index = 0, set = 0, binding = 0) uniform subpassInput inputColor;
```

_from what I understand on why specification defines things this way:_
1. Framebuffers are sometimes inside special memory regions. In this scenario, a input_attachment_index can 'directly' fetch a pointer to the memory.
2. Sometimes hardware stores a framebuffer as a regular image, this means to figure out where the image is in memory a descriptor can be used.

Reply:
It starts to sound to me like the variable `inputColor` should have two `ParameterBindingInfo` entries, one for DescriptorTableSlot, and one for InputAttachment.

Reply:
In `addExplicitParameterBindings_GLSL`, we have the opportunity to insert as many `ParameterBindingInfo` as we want based on the modifiers on the variable.